### PR TITLE
SLiM population size trajectory adjustment

### LINKED ISSUE #1589: SLiM engine truncates the number of diploids to match generation 1
I am simulating the `OutOfAfrica_3G09` model, which suggests that the maximum population size of the CEU population at time 0 is 29725 diploids:

```python
import stdpopsim
species = stdpopsim.get_species("HomSap")
model = species.get_demographic_model("OutOfAfrica_3G09")
for pop, size in zip(model.populations, model.model.debug().population_size_trajectory([0])[0]):
    print("At t=0", pop.name, "is predicted to be of size", size)

> At t=0 YRI is predicted to be of size 12300.0
> At t=0 CEU is predicted to be of size 29725.343546388514
> At t=0 CHB is predicted to be of size 54090.331077946525
```

However, if I ask for 29640 CEU samples at time 0, I only get back 29607 diploids:

```python
samples = {'YRI': 1248, 'CHB': 312, 'CEU': 29640}
engine = stdpopsim.get_engine("slim")
contig = species.get_contig("chr17")
ts = engine.simulate(model, contig, samples, seed=123, slim_burn_in=10)  # takes a few hours
print(len(ts.samples(population=[p.id for p in ts.populations() if p.metadata['name']=="CEU"][0]))//2, "CEU diploids")

>>> 29607 CEU diploids
```

This number is suspiciously close to the number of individuals predicted in generation 1:

```python
for pop, size in zip(model.populations, model.model.debug().population_size_trajectory([1])[0]):
    print("At t=1", pop.name, "is predicted to be of size", size)
> At t=1 YRI is predicted to be of size 12300.0
> At t=1 CEU is predicted to be of size 29606.679658197816
> At t=1 CHB is predicted to be of size 53793.650875455634
```

I thought that SLiM warns when it can't create all the requested samples, but I can't see anything in the error output.

### PR DESCRIPTION
**Pull request by petrelharp**
We did not have robust checks for SLiM's population size trajectories. That's a good bit better now (although we could probably think of more complicated demographic scenarios to check). Issues I uncovered:
- we were effectively computing population sizes during growth phases by doing
   `round(round(exp(-r*t)*final_size) * exp(r*t))`; since `round` and `exp` don't commute,
  this is not ideal; so I removed the inner `round` (storing initial sizes as floats not ints)
- since population sizes are set in `late()` but these don't take effect until the next time step, the final population sizes was not what you'd think (it was more like what it would have been in the previous time step); I've changed the growth rate used by SLiM slightly so that they agree. A better fix might be to do all that stuff in `early()` but that's a giant can of worms with unknown consequences in everything else, so not going there.

There's definitely the possibility that I've majorly messed something up, so someone should have a very careful review of this.

Closes #1589.

### CODE REVIEW

File: tests/test_slim_engine.py [hunk 1]
```diff
@@ -3079,6 +3079,79 @@ def test_errors(self):
             stdpopsim.selection_coeff_from_mutation(ts, "bar")
 
 
+@pytest.mark.skipif(IS_WINDOWS, reason="SLiM not available on windows")
```
Comment (bhaller) [inline tests/test_slim_engine.py:3082]: I don't know the stdpopsim code at all, but just here kibitzing, to say: SLiM *is* available on Windows... :->

File: tests/test_slim_engine.py [hunk 2]
```diff
@@ -3079,6 +3079,79 @@ def test_errors(self):
...
+    time t we set the size to x; then subsequently we re-set the size
+    only T-1 times. (One consideration is that we re-set sizes in `late`,
+    which doesn't take effect until the next tick, so a reset in the final
+    time step has no effect.) This means that the continuous-time msprime model
+    that we base this on is growing for total time T, but the SLiM model grows
+    in T-1 time steps. Furthermore, we'd like both the final and initial sizes
+    to be exactly what they are supposed to be. This amounts to approximating
+    exp(r*t) over the interval [0,T] by exp(r*t*T/(T-1)). Probably something
+    nicer could be done, but this is close. The approximation is better for
+    smaller growth rates.
```
Comment (bhaller) [inline tests/test_slim_engine.py:3098]: It won't tell you anything you don't already know, but note that the SLiM manual, section 5.1.2, has a extended (probably too extended!) discussion of exactly these sorts of issues for exponential growth.  Possibly there would be something useful for you there.

Comment (bhaller) [reply tests/test_slim_engine.py:3098]: or possibly you will tell me that I should rewrite that section in a less dumb/boring way, which is also fine :->

File: tests/test_slim_engine.py (hunk 1 above)
Comment (petrelharp) [reply tests/test_slim_engine.py:3082]: That's a different issue that is open still because getting temp files to work correctly in Windows is a Pain.

**Comment by nspope** [comment]
I've looked over the logic here, and it all *looks* correct for the scaling factor. As an extra verification, I simulated from the zigzag model and calculated pair coalescence rates ("empirical" population sizes) for (1) msprime, (2) slim, (3) slim with scaling factor of 2. Here's what that looks like:



<details>
<summary>code for zigzag</summary>

```python
import stdpopsim
import tskit
import numpy as np
import os

overwrite = False
sequence_length = 50e6
samples = 1000

homsap = stdpopsim.get_species("HomSap")
zigzag = homsap.get_demographic_model("Zigzag_1S14")
contig = homsap.get_contig(length=sequence_length)
sample = {"generic" : samples}

msp_engine = stdpopsim.get_engine("msprime")
slim_engine = stdpopsim.get_engine("slim")


msp_trees = "zigzag.msprime.trees"
if not os.path.exists(msp_trees) or overwrite:
    msp_ts = msp_engine.simulate(
        contig=contig, 
        demographic_model=zigzag, 
        samples=sample, 
        seed=1234,
    )
    msp_ts.dump(msp_trees)
else:
    msp_ts = tskit.load(msp_trees)


slim_trees = "zigzag.slim.trees"
if not os.path.exists(slim_trees) or overwrite:
    slim_ts = slim_engine.simulate(
        contig=contig, 
        demographic_model=zigzag, 
        samples=sample, 
        slim_burn_in=1,
        verbosity=3,
        seed=1234,
    )
    slim_ts.dump(slim_trees)
else:
    slim_ts = tskit.load(slim_trees)


slimscale_trees = "zigzag.slimscale.trees"
if not os.path.exists(slimscale_trees) or overwrite:
    slimscale_ts = slim_engine.simulate(
        contig=contig, 
        demographic_model=zigzag, 
        samples=sample, 
        slim_burn_in=1,
        slim_scaling_factor=2,
        verbosity=3,
        seed=1234,
    )
    slimscale_ts.dump(slimscale_trees)
else:
    slimscale_ts = tskit.load(slimscale_trees)


time_grid = np.logspace(2, 5, 51)
time_grid[0] = 0.0
time_grid[-1] = np.inf

msp_rates = msp_ts.pair_coalescence_rates(time_windows=time_grid)
slim_rates = slim_ts.pair_coalescence_rates(time_windows=time_grid)
slimscale_rates = slimscale_ts.pair_coalescence_rates(time_windows=time_grid)

import matplotlib.pyplot as plt
midpoint = (time_grid[1:] + time_grid[:-1]) / 2
plt.step(midpoint, msp_rates, where="mid", label="msprime", color="red")
plt.step(midpoint, slim_rates, where="mid", label="slim", color="blue")
plt.step(midpoint, slimscale_rates, where="mid", label="slim-scale2", color="green")
plt.xscale("log")
plt.xlabel("Generations in past")
plt.ylabel("Pair coalescence rate")
plt.title("Zigzag_1S14")
plt.legend()
plt.savefig("/home/natep/public_html/stdpopsim/slim-pop-size-review/zigzag.coal_rates.png")
```

</details>

which looks fine (and has lots of exponential growth/decline). I did the same thing for one of the out-of-Africa models, splitting by population, which also serves as a check for how population splits/migration rates are handled:



<details>
<summary>code for ooa</summary>

```python
import stdpopsim
import numpy as np
import tskit
import os

overwrite = False
sequence_length = 50e6
samples = 1000
prefix = "ooa"

homsap = stdpopsim.get_species("HomSap")
zigzag = homsap.get_demographic_model("OutOfAfrica_4J17")
contig = homsap.get_contig(length=sequence_length)
sample = {"CHB" : samples, "CEU" : samples, "YRI" : samples, "JPT" : samples}

msp_engine = stdpopsim.get_engine("msprime")
slim_engine = stdpopsim.get_engine("slim")


msp_trees = f"{prefix}.msprime.trees"
if not os.path.exists(msp_trees) or overwrite:
    msp_ts = msp_engine.simulate(
        contig=contig, 
        demographic_model=zigzag, 
        samples=sample, 
        seed=1234,
    )
    msp_ts.dump(msp_trees)
else:
    msp_ts = tskit.load(msp_trees)


slim_trees = f"{prefix}.slim.trees"
if not os.path.exists(slim_trees) or overwrite:
    slim_ts = slim_engine.simulate(
        contig=contig, 
        demographic_model=zigzag, 
        samples=sample, 
        slim_burn_in=1,
        verbosity=3,
        seed=1234,
    )
    slim_ts.dump(slim_trees)
else:
    slim_ts = tskit.load(slim_trees)


slimscale_trees = f"{prefix}.slimscale.trees"
if not os.path.exists(slimscale_trees) or overwrite:
    slimscale_ts = slim_engine.simulate(
        contig=contig, 
        demographic_model=zigzag, 
        samples=sample, 
        slim_burn_in=1,
        slim_scaling_factor=2,
        verbosity=3,
        seed=1234,
    )
    slimscale_ts.dump(slimscale_trees)
else:
    slimscale_ts = tskit.load(slimscale_trees)


def extract_sample_sets(ts):
    population_map = {
        i:p.metadata['name'] for i, p 
        in enumerate(ts.populations())
    }
    sample_sets = {}
    for i, p in population_map.items():
        in_pop = np.logical_and(
            np.bitwise_and(ts.nodes_flags, tskit.NODE_IS_SAMPLE), 
            ts.nodes_population == i,
        )
        sample_sets[p] = np.flatnonzero(in_pop)
    return [sample_sets["CEU"], sample_sets["CHB"], sample_sets["JPT"], sample_sets["YRI"]]


time_grid = np.logspace(2, 5, 51)
time_grid[0] = 0.0
time_grid[-1] = np.inf

msp_rates = msp_ts.pair_coalescence_rates(
    time_windows=time_grid, 
    sample_sets=extract_sample_sets(msp_ts), 
    indexes=[[0,0],[1,1],[2,2],[3,3]]
)
slim_rates = slim_ts.pair_coalescence_rates(
    time_windows=time_grid, 
    sample_sets=extract_sample_sets(slim_ts), 
    indexes=[[0,0],[1,1],[2,2],[3,3]]
)
slimscale_rates = slimscale_ts.pair_coalescence_rates(
    time_windows=time_grid, 
    sample_sets=extract_sample_sets(slimscale_ts), 
    indexes=[[0,0],[1,1],[2,2],[3,3]]
)

import matplotlib.pyplot as plt
midpoint = (time_grid[1:] + time_grid[:-1]) / 2
pop_names = ["CEU", "CHB", "JPT", "YRI"]
fig, axs = plt.subplots(2, 2, figsize=(8, 8), constrained_layout=True)
for i, ax in enumerate(np.ravel(axs)):
    ax.step(midpoint, msp_rates[i], where="mid", label="msprime", color="red")
    ax.step(midpoint, slim_rates[i], where="mid", label="slim", color="blue")
    ax.step(midpoint, slimscale_rates[i], where="mid", label="slim-scale2", color="green")
    ax.legend()
    ax.set_title(f"{pop_names[i]}")
    ax.set_xscale("log")
fig.supxlabel("Generations in past")
fig.supylabel("Pair coalescence rate")
fig.suptitle("OutOfAfrica_4J17")
plt.savefig("/home/natep/public_html/stdpopsim/slim-pop-size-review/ooa.coal_rates.png")
```

</details>

So, everything seems to be working as expected here. I'm going to take another look at the tests (tomorrow) to be sure.

**Comment by petrelharp** [comment]
I've got a test for rescaling! It's a bit icky - the alignment in time seems to be +/-1 - but I'm good saying it's due to Discretization Issues and calling it good?

File: tests/test_slim_engine.py (hunk 2 above)
Comment (petrelharp) [reply tests/test_slim_engine.py:3098]: I just learned some things from that! 
Screenshot from 2024-12-18 21-16-05

File: tests/test_slim_engine.py [hunk 3]
```diff
@@ -3079,6 +3079,100 @@ def test_errors(self):
...
+        for pop in slim_sizes:
+            k = [x.name for x in model.populations].index(pop)
+            pop_t = np.array(slim_sizes[pop][0]["t"])
+            pop_N = np.array(slim_sizes[pop][0]["N"])
+            dd_N = dd.population_size_trajectory(np.floor((T - pop_t) * Q))[:, k] / Q
+            dd_N1 = (
+                dd.population_size_trajectory(np.floor((T - pop_t + 1) * Q))[:, k] / Q
+            )
+            # np.testing.assert_allclose(pop_N, dd_N, rtol=0.01)
+            assert np.all(
```
Comment (nspope) [inline tests/test_slim_engine.py:3133]: Worth leaving a note saying the double check here is to deal with discretization error

File: stdpopsim/slim_engine.py [hunk 4]
```diff
@@ -932,14 +959,14 @@ def fix_time(event):
...
-    N = np.empty(shape=(dd.num_populations, len(epochs)), dtype=int)
+    N = np.empty(shape=(dd.num_populations, len(epochs)), dtype=float)
     growth_rates = np.empty(shape=(dd.num_populations, len(epochs)), dtype=float)
     for j, epoch in enumerate(epochs):
         for i, pop in enumerate(epoch.populations):
             # SLiM simulates a diploid population, so rescale population size
             # depending on contig ploidy. If/when the SLiM simulation is
             # converted to a haploid model, this rescaling should be removed.
-            N[i, j] = int(pop.end_size * contig.ploidy / 2)
+            N[i, j] = pop.end_size * contig.ploidy / 2
```
Comment (nspope) [inline stdpopsim/slim_engine.py:969]: This should be getting covered by the tests, but codecov disagrees ... ?

**Review by nspope — APPROVED** [review]
LGTM -- could you update the changelog?

File: stdpopsim/slim_engine.py (hunk 4 above)
Comment (nspope) [reply stdpopsim/slim_engine.py:969]: Nevermind-- codecov itself shows these as covered, the github API isn't picking up on that for some reason.

**Pull request merged**